parsel/selector.py

```diff
@@ -90,9 +90,15 @@ def create_root_node(
     parser_cls: Type[_ParserType],
     base_url: Optional[str] = None,
     huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
+    body: Optional[bytes] = "",
+    encoding: Optional[str] = "utf8",
 ) -> etree._Element:
     """Create root node for text using given parser class."""
-    body = text.strip().replace("\x00", "").encode("utf8") or b"<html/>"
+    if body is not None and not text:
```

Comment:
We can simplify this if we expect body to always be bytes.

```suggestion
    if not text:
```

Reply:
Changing this conditon caused several tests to fail.
<details><summary>tests log output fragmenr</summary>

```
tests\test_selector.py:836 (SelectorTestCase.test_empty_bodies_shouldnt_raise_errors)
self = <test_selector.SelectorTestCase testMethod=test_empty_bodies_shouldnt_raise_errors>

    def test_empty_bodies_shouldnt_raise_errors(self) -> None:
>       self.sscls(text="").xpath("//text()").extract()

test_selector.py:838: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\parsel\selector.py:375: in __init__
    root = self._get_root(text, base_url, huge_tree)
..\parsel\selector.py:401: in _get_root
    return create_root_node(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

text = '', parser_cls = <class 'lxml.html.HTMLParser'>, base_url = None
huge_tree = True, body = None, encoding = 'utf8'

    def create_root_node(
        text: str,
        parser_cls: Type[_ParserType],
        base_url: Optional[str] = None,
        huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
        body: bytes = "",
        encoding: str = "utf8",
    ) -> etree._Element:
        """Create root node for text using given parser class."""
        if not text:
>           body = body.replace(b"\x00", b"").strip()
E           AttributeError: 'NoneType' object has no attribute 'replace'

..\parsel\selector.py:98: AttributeError
PASSED [  8%]PASSED       [  8%]PASSED            [  9%]PASSED    [ 10%]PASSED [ 11%]PASSED [ 11%]PASSED            [ 12%]PASSED    [ 13%]PASSED       [ 14%]PASSED      [ 14%]PASSED   [ 15%]PASSED         [ 16%]PASSED [ 17%]PASSED      [ 17%]PASSED [ 18%]PASSED        [ 19%]SKIPPED (Text nodes lost parent node reference in lxml) [ 20%]
Skipped: Text nodes lost parent node reference in lxml
PASSED         [ 20%]PASSED [ 21%]PASSED          [ 22%]PASSED     [ 23%]PASSED                       [ 23%]PASSED                 [ 24%]PASSED                  [ 25%]PASSED      [ 26%]PASSED [ 26%]PASSED        [ 27%]PASSED [ 28%]PASSED [ 29%]PASSED [ 29%]PASSED [ 30%]PASSED          [ 31%]PASSED     [ 32%]PASSED [ 32%]PASSED [ 33%]PASSED     [ 34%]PASSED [ 35%]PASSED     [ 35%]PASSED [ 36%]PASSED     [ 37%]PASSED       [ 38%]PASSED    [ 38%]PASSED       [ 39%]PASSED   [ 40%]PASSED [ 41%]FAILED         [ 41%]
tests\test_selector.py:52 (SelectorTestCase.test_simple_selection)
self = <test_selector.SelectorTestCase testMethod=test_simple_selection>

    def test_simple_selection(self) -> None:
        """Simple selector tests"""
        body = "<p><input name='a'value='1'/><input name='b'value='2'/></p>"
        sel = self.sscls(text=body)
    
        xl = sel.xpath("//input")
        self.assertEqual(2, len(xl))
        for x in xl:
>           self.assertIsSelector(x)

test_selector.py:61: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_selector.py:27: in assertIsSelector
    self.assertEqual(type(value), type(self.sscls(text="")))
..\parsel\selector.py:375: in __init__
    root = self._get_root(text, base_url, huge_tree)
..\parsel\selector.py:401: in _get_root
    return create_root_node(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

text = '', parser_cls = <class 'lxml.html.HTMLParser'>, base_url = None
huge_tree = True, body = None, encoding = 'utf8'

    def create_root_node(
        text: str,
        parser_cls: Type[_ParserType],
        base_url: Optional[str] = None,
        huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
        body: bytes = "",
        encoding: str = "utf8",
    ) -> etree._Element:
        """Create root node for text using given parser class."""
        if not text:
>           body = body.replace(b"\x00", b"").strip()
E           AttributeError: 'NoneType' object has no attribute 'replace'

..\parsel\selector.py:98: AttributeError
PASSED [ 42%]PASSED [ 43%]FAILED                  [ 44%]
tests\test_selector.py:438 (SelectorTestCase.test_slicing)
self = <test_selector.SelectorTestCase testMethod=test_slicing>

    def test_slicing(self) -> None:
        text = "<div><p>1</p><p>2</p><p>3</p></div>"
        hs = self.sscls(text=text, type="html")
>       self.assertIsSelector(hs.css("p")[2])

test_selector.py:442: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_selector.py:27: in assertIsSelector
    self.assertEqual(type(value), type(self.sscls(text="")))
..\parsel\selector.py:375: in __init__
    root = self._get_root(text, base_url, huge_tree)
..\parsel\selector.py:401: in _get_root
    return create_root_node(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

text = '', parser_cls = <class 'lxml.html.HTMLParser'>, base_url = None
huge_tree = True, body = None, encoding = 'utf8'

    def create_root_node(
        text: str,
        parser_cls: Type[_ParserType],
        base_url: Optional[str] = None,
        huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
        body: bytes = "",
        encoding: str = "utf8",
    ) -> etree._Element:
        """Create root node for text using given parser class."""
        if not text:
>           body = body.replace(b"\x00", b"").strip()
E           AttributeError: 'NoneType' object has no attribute 'replace'

..\parsel\selector.py:98: AttributeError
PASSED            [ 44%]PASSED [ 45%]FAILED            [ 46%]
tests\test_selector.py:891 (SelectorTestCase.test_weakref_slots)
self = <test_selector.SelectorTestCase testMethod=test_weakref_slots>

    def test_weakref_slots(self) -> None:
        """Check that classes are using slots and are weak-referenceable"""
>       x = self.sscls(text="")

test_selector.py:894: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\parsel\selector.py:375: in __init__
    root = self._get_root(text, base_url, huge_tree)
..\parsel\selector.py:401: in _get_root
    return create_root_node(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

text = '', parser_cls = <class 'lxml.html.HTMLParser'>, base_url = None
huge_tree = True, body = None, encoding = 'utf8'

    def create_root_node(
        text: str,
        parser_cls: Type[_ParserType],
        base_url: Optional[str] = None,
        huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
        body: bytes = "",
        encoding: str = "utf8",
    ) -> etree._Element:
        """Create root node for text using given parser class."""
        if not text:
>           body = body.replace(b"\x00", b"").strip()
E           AttributeError: 'NoneType' object has no attribute 'replace'

..\parsel\selector.py:98: AttributeError
```

</details>

Reply:
@Gallaecio I`ve applied all other updates in code. Changing this condition causing some tests to fail.

Reply:
That failure seems caused by `body` being `None`, which should never be the case. I think it stems from the fact that you have not make these changes also in the `Selector.__init__` method.

Reply:
Fixed. (updated in `_get_root`)